File: src/main/java/com/davidm1a2/afraidofthedark/common/entity/mcAnimatorLib/animation/AnimationHandler.java
package com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.animation;

import com.davidm1a2.afraidofthedark.AfraidOfTheDark;
import com.davidm1a2.afraidofthedark.client.entity.mcAnimatorLib.MCAModelRenderer;
import com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.IMCAnimatedModel;
import com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.math.Quaternion;
import com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.math.Vector3f;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.*;

public abstract class AnimationHandler {
    /**
     * List of all the activate animations of this Entity.
     */
    public List<Channel> animCurrentChannels = new ArrayList<Channel>();
    /**
     * Previous time of every active animation.
     */
    public HashMap<String, Long> animPrevTime = new HashMap<String, Long>();
    /**
     * Current frame of every active animation.
     */
    public HashMap<String, Float> animCurrentFrame = new HashMap<String, Float>();
    /**
     * Owner of this handler.
     */
    private IMCAnimatedModel animatedModel;

    public AnimationHandler(IMCAnimatedModel model) {
        //AfraidOfTheDark.INSTANCE.getAnimationHandler().addEntity(entity);
        animatedModel = model;
    }

    /**
     * Update animation values. Return false if the animation should stop.
     */
    public static boolean updateAnimation(IMCAnimatedModel model, Channel channel, HashMap<String, Long> prevTimeAnim, HashMap<String, Float> prevFrameAnim) {
        if (FMLCommonHandler.instance().getEffectiveSide().isServer() || (FMLCommonHandler.instance().getEffectiveSide().isClient() && !isGamePaused())) {
            if (!(channel.mode == Channel.CUSTOM)) {
                long prevTime = prevTimeAnim.get(channel.name);
                float prevFrame = prevFrameAnim.get(channel.name);

                long currentTime = System.nanoTime();
                double deltaTime = (currentTime - prevTime) / 1000000000.0;
                float numberOfSkippedFrames = (float) (deltaTime * channel.fps);

                float currentFrame = prevFrame + numberOfSkippedFrames;

                if (currentFrame < channel.totalFrames - 1) //-1 as the first frame mustn't be "executed" as it is the starting situation
                {
                    prevTimeAnim.put(channel.name, currentTime);
                    prevFrameAnim.put(channel.name, currentFrame);
                    return true;
                } else {
                    if (channel.mode == Channel.LOOP) {
                        prevTimeAnim.put(channel.name, currentTime);
                        prevFrameAnim.put(channel.name, 0F);
                        return true;
                    }
                    return false;
                }
            } else {
                return true;
            }
        } else {
            long currentTime = System.nanoTime();
            prevTimeAnim.put(channel.name, currentTime);
            return true;
        }
    }

    @SideOnly(Side.CLIENT)
    private static boolean isGamePaused() {
        net.minecraft.client.Minecraft MC = net.minecraft.client.Minecraft.getMinecraft();
        return MC.isSingleplayer() && MC.currentScreen != null && MC.currentScreen.doesGuiPauseGame() && !MC.getIntegratedServer().getPublic();
    }

    /**
     * Apply animations if running or apply initial values. Must be called only by the model class.
     */
    @SideOnly(Side.CLIENT)
    public static void performAnimationInModel(Map<String, MCAModelRenderer> parts, IMCAnimatedModel model) {
        for (Map.Entry<String, MCAModelRenderer> entry : parts.entrySet()) {
            String boxName = entry.getKey();
            MCAModelRenderer box = entry.getValue();

            boolean anyRotationApplied = false;
            boolean anyTranslationApplied = false;
            boolean anyCustomAnimationRunning = false;

            for (Channel channel : model.getAnimationHandler().animCurrentChannels) {
                if (channel.mode != Channel.CUSTOM) {
                    float currentFrame = model.getAnimationHandler().animCurrentFrame.get(channel.name);

                    //Rotations
                    KeyFrame prevRotationKeyFrame = channel.getPreviousRotationKeyFrameForBox(boxName, model.getAnimationHandler().animCurrentFrame.get(channel.name));
                    int prevRotationKeyFramePosition = prevRotationKeyFrame != null ? channel.getKeyFramePosition(prevRotationKeyFrame) : 0;

                    KeyFrame nextRotationKeyFrame = channel.getNextRotationKeyFrameForBox(boxName, model.getAnimationHandler().animCurrentFrame.get(channel.name));
                    int nextRotationKeyFramePosition = nextRotationKeyFrame != null ? channel.getKeyFramePosition(nextRotationKeyFrame) : 0;

                    float SLERPProgress = (currentFrame - prevRotationKeyFramePosition) / (nextRotationKeyFramePosition - prevRotationKeyFramePosition);
                    if (SLERPProgress > 1F || SLERPProgress < 0F) {
                        SLERPProgress = 1F;
                    }

                    if (prevRotationKeyFramePosition == 0 && prevRotationKeyFrame == null && !(nextRotationKeyFramePosition == 0)) {
                        Quaternion currentQuat = new Quaternion();
                        currentQuat.slerp(parts.get(boxName).getDefaultRotationAsQuaternion(), nextRotationKeyFrame.modelRenderersRotations.get(boxName), SLERPProgress);
                        box.getRotationMatrix().set(currentQuat).transpose();

                        anyRotationApplied = true;
                    } else if (prevRotationKeyFramePosition == 0 && prevRotationKeyFrame != null && !(nextRotationKeyFramePosition == 0)) {
                        Quaternion currentQuat = new Quaternion();
                        currentQuat.slerp(prevRotationKeyFrame.modelRenderersRotations.get(boxName), nextRotationKeyFrame.modelRenderersRotations.get(boxName), SLERPProgress);
                        box.getRotationMatrix().set(currentQuat).transpose();

                        anyRotationApplied = true;
                    } else if (!(prevRotationKeyFramePosition == 0) && !(nextRotationKeyFramePosition == 0)) {
                        Quaternion currentQuat = new Quaternion();
                        currentQuat.slerp(prevRotationKeyFrame.modelRenderersRotations.get(boxName), nextRotationKeyFrame.modelRenderersRotations.get(boxName), SLERPProgress);
                        box.getRotationMatrix().set(currentQuat).transpose();

                        anyRotationApplied = true;
                    }

                    //Translations
                    KeyFrame prevTranslationKeyFrame = channel.getPreviousTranslationKeyFrameForBox(boxName, model.getAnimationHandler().animCurrentFrame.get(channel.name));
                    int prevTranslationsKeyFramePosition = prevTranslationKeyFrame != null ? channel.getKeyFramePosition(prevTranslationKeyFrame) : 0;

                    KeyFrame nextTranslationKeyFrame = channel.getNextTranslationKeyFrameForBox(boxName, model.getAnimationHandler().animCurrentFrame.get(channel.name));
                    int nextTranslationsKeyFramePosition = nextTranslationKeyFrame != null ? channel.getKeyFramePosition(nextTranslationKeyFrame) : 0;

                    float LERPProgress = (currentFrame - prevTranslationsKeyFramePosition) / (nextTranslationsKeyFramePosition - prevTranslationsKeyFramePosition);
                    if (LERPProgress > 1F) {
                        LERPProgress = 1F;
                    }

                    if (prevTranslationsKeyFramePosition == 0 && prevTranslationKeyFrame == null && !(nextTranslationsKeyFramePosition == 0)) {
                        Vector3f startPosition = parts.get(boxName).getPositionAsVector();
                        Vector3f endPosition = nextTranslationKeyFrame.modelRenderersTranslations.get(boxName);
                        Vector3f currentPosition = new Vector3f(startPosition);
                        currentPosition.interpolate(endPosition, LERPProgress);
                        box.setRotationPoint(currentPosition.x, currentPosition.y, currentPosition.z);

                        anyTranslationApplied = true;
                    } else if (prevTranslationsKeyFramePosition == 0 && prevTranslationKeyFrame != null && !(nextTranslationsKeyFramePosition == 0)) {
                        Vector3f startPosition = prevTranslationKeyFrame.modelRenderersTranslations.get(boxName);
                        Vector3f endPosition = nextTranslationKeyFrame.modelRenderersTranslations.get(boxName);
                        Vector3f currentPosition = new Vector3f(startPosition);
                        currentPosition.interpolate(endPosition, LERPProgress);
                        box.setRotationPoint(currentPosition.x, currentPosition.y, currentPosition.z);
                    } else if (!(prevTranslationsKeyFramePosition == 0) && !(nextTranslationsKeyFramePosition == 0)) {
                        Vector3f startPosition = prevTranslationKeyFrame.modelRenderersTranslations.get(boxName);
                        Vector3f endPosition = nextTranslationKeyFrame.modelRenderersTranslations.get(boxName);
                        Vector3f currentPosition = new Vector3f(startPosition);
                        currentPosition.interpolate(endPosition, LERPProgress);
                        box.setRotationPoint(currentPosition.x, currentPosition.y, currentPosition.z);

                        anyTranslationApplied = true;
                    }
                } else {
                    anyCustomAnimationRunning = true;

                    ((CustomChannel) channel).update(parts, model);
                }
            }

            //Set the initial values for each box if necessary
            if (!anyRotationApplied && !anyCustomAnimationRunning) {
                box.resetRotationMatrix();
            }
            if (!anyTranslationApplied && !anyCustomAnimationRunning) {
                box.resetRotationPoint();
            }
        }
    }

    public IMCAnimatedModel getEntity() {
        return animatedModel;
    }

    public void activateAnimation(Map<String, Channel> animChannels, String name, float startingFrame) {
        if (animChannels.get(name) != null) {
            Channel selectedChannel = animChannels.get(name);
            int indexToRemove = animCurrentChannels.indexOf(selectedChannel);
            if (indexToRemove != -1) {
                animCurrentChannels.remove(indexToRemove);
            }

            animCurrentChannels.add(selectedChannel);
            animPrevTime.put(name, System.nanoTime());
            animCurrentFrame.put(name, startingFrame);
        } else {
            System.out.println("The animation called " + name + " doesn't exist!");
        }
    }

    public abstract void activateAnimation(String name, float startingFrame);

    public void stopAnimation(Map<String, Channel> animChannels, String name) {
        Channel selectedChannel = animChannels.get(name);
        if (selectedChannel != null) {
            int indexToRemove = animCurrentChannels.indexOf(selectedChannel);
            if (indexToRemove != -1) {
                animCurrentChannels.remove(indexToRemove);
                animPrevTime.remove(name);
                animCurrentFrame.remove(name);
            }
        } else {
            System.out.println("The animation called " + name + " doesn't exist!");
        }
    }

    public abstract void stopAnimation(String name);

    public void animationsUpdate() {
        ListIterator<Channel> it = animCurrentChannels.listIterator();
        while (it.hasNext()) {
            Channel anim = it.next();
            if (animCurrentFrame == null) {
                AfraidOfTheDark.INSTANCE.getLogger().info("animCurrentFrame = null");
            }
            if (anim == null) {
                AfraidOfTheDark.INSTANCE.getLogger().info("anim = null");
            }
            float prevFrame = animCurrentFrame.get(anim.name);
            boolean animStatus = updateAnimation(animatedModel, anim, animPrevTime, animCurrentFrame);

            if (!animStatus) {
                //channelsToRemove.add(anim);
                animPrevTime.remove(anim.name);
                animCurrentFrame.remove(anim.name);
                it.remove();
            }
        }

        //		/*
        //		 * Added this
        //		 */
        //		for (Channel channel : channelsToRemove)
        //		{
        //			animCurrentChannels.remove(channel);
        //			animPrevTime.remove(channel.name);
        //			animCurrentFrame.remove(channel.name);
        //			animationEvents.get(channel.name).clear();
        //		}
        //
        //		channelsToRemove.clear();
    }

    public boolean isAnimationActive(String name) {
        for (Channel anim : animatedModel.getAnimationHandler().animCurrentChannels) {
            if (anim.name.equals(name)) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/kotlin/com/davidm1a2/afraidofthedark/common/entity/mcAnimatorLib/animation/AnimationHandler.kt
package com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.animation

import com.davidm1a2.afraidofthedark.client.entity.mcAnimatorLib.MCAModelRenderer
import com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.interpolate
import com.davidm1a2.afraidofthedark.common.entity.mcAnimatorLib.slerp
import net.minecraft.client.Minecraft
import net.minecraft.util.math.vector.Quaternion
import net.minecraftforge.fml.LogicalSide
import net.minecraftforge.fml.common.thread.EffectiveSide

/**
 * Constructor takes in all available animation channels
 *
 * @param animChannels A list of channels this animation handler must support
 * @property availableAnimChannels A map of name to available channels
 * @property animCurrentChannels List of all the activate animations of this Entity
 * @property animPrevTime Previous time of every active animation
 * @property animCurrentFrame Current frame of every active animation
 */
class AnimationHandler(vararg animChannels: Channel) {
    private val animCurrentChannels = mutableSetOf<Channel>()
    private val animPrevTime = mutableMapOf<String, Long>()
    private val animCurrentFrame = mutableMapOf<String, Float>()
    private val availableAnimChannels = animChannels.associateBy { it.name }

    /**
     * Performs one "tick" of the animation
     */
    fun update() {
        // Go over every channel, update it, and remove it if it's finished
        val it = animCurrentChannels.iterator()
        while (it.hasNext()) {
            val anim = it.next()
            val animStatus = updateAnimation(anim)
            if (!animStatus) {
                animPrevTime.remove(anim.name)
                animCurrentFrame.remove(anim.name)
                it.remove()
            }
        }
    }

    /**
     * Plays a given animation from a given frame
     *
     * @param name The animation to play
     * @param startingFrame The frame to start the animation at
     */
    fun playAnimation(name: String, startingFrame: Float = 0f) {
        val channel = availableAnimChannels[name]
        if (channel != null) {
            animCurrentChannels.add(channel)
            animPrevTime[name] = System.nanoTime()
            animCurrentFrame[name] = startingFrame
        } else {
            println("The animation called $name doesn't exist!")
        }
    }

    /**
     * Stops a given animation
     *
     * @param name The animation to stop
     */
    fun stopAnimation(name: String) {
        val channel = availableAnimChannels[name]
        if (channel != null) {
            if (animCurrentChannels.contains(channel)) {
                animCurrentChannels.remove(channel)
                animPrevTime.remove(name)
                animCurrentFrame.remove(name)
            }
        } else {
            println("The animation called $name doesn't exist!")
        }
    }

    /**
     * Tests if a given animation is active
     *
     * @param name The animation to test
     * @return True if the animation is playing, false otherwise
     */
    fun isAnimationActive(name: String): Boolean {
        return animCurrentChannels.any { it.name == name }
    }

    /**
     * Update animation values
     *
     * @param channel The channel to update
     * @return false if the animation should stop, true otherwise
     */
    private fun updateAnimation(channel: Channel): Boolean {
        val side = EffectiveSide.get()
        return if (side == LogicalSide.SERVER || side == LogicalSide.CLIENT && !isGamePaused()) {
            if (channel.mode != ChannelMode.CUSTOM) {
                val prevTime = animPrevTime[channel.name]!!
                val prevFrame = animCurrentFrame[channel.name]!!
                val currentTime = System.nanoTime()
                val deltaTime = (currentTime - prevTime) / 1000000000.0
                val numberOfSkippedFrames = (deltaTime * channel.fps).toFloat()
                val currentFrame = prevFrame + numberOfSkippedFrames
                if (currentFrame < channel.totalFrames) {
                    animPrevTime[channel.name] = currentTime
                    animCurrentFrame[channel.name] = currentFrame
                    true
                } else {
                    if (channel.mode == ChannelMode.LOOP) {
                        animPrevTime[channel.name] = currentTime
                        animCurrentFrame[channel.name] = 0f
                        return true
                    }
                    false
                }
            } else {
                true
            }
        } else {
            val currentTime = System.nanoTime()
            animPrevTime[channel.name] = currentTime
            true
        }
    }

    /**
     * @return true if the game is paused, false if not
     */
    private fun isGamePaused(): Boolean {
        val minecraft = Minecraft.getInstance()
        return minecraft.hasSingleplayerServer() && minecraft.screen != null && minecraft.screen!!.isPauseScreen && !minecraft.singleplayerServer!!.isSingleplayer
    }

    /**
     * Apply animations if running or apply initial values. Must be called only by the model class
     *
     * @param parts The parts of the model to update
     */
    fun performAnimationInModel(parts: Map<String, MCAModelRenderer>) {
        for ((boxName, box) in parts) {
            for (channel in animCurrentChannels) {
                if (channel.mode != ChannelMode.CUSTOM) {
                    val currentFrame = animCurrentFrame[channel.name]!!

                    // Rotations
                    val (prevRotationKeyFramePosition, prevRotationKeyFrame) =
                        channel.getPreviousRotationKeyFrameForBox(boxName, animCurrentFrame[channel.name]!!)
                    val (nextRotationKeyFramePosition, nextRotationKeyFrame) =
                        channel.getNextRotationKeyFrameForBox(boxName, animCurrentFrame[channel.name]!!)

                    var slerpProgress =
                        (currentFrame - prevRotationKeyFramePosition) / (nextRotationKeyFramePosition - prevRotationKeyFramePosition)
                    slerpProgress = slerpProgress.coerceIn(0f, 1f)

                    if (prevRotationKeyFrame == null && nextRotationKeyFrame != null) { // Before the first key frame
                        val currentQuat = Quaternion(0f, 0f, 0f, 0f)
                        currentQuat.slerp(
                            box.defaultRotation,
                            nextRotationKeyFrame.modelRotations[boxName]!!,
                            slerpProgress
                        )
                        box.rotation.set(currentQuat.i(), currentQuat.j(), currentQuat.k(), currentQuat.r())
                    } else if (prevRotationKeyFrame != null && nextRotationKeyFrame == null) { // After the last keyframe
                        val currentQuat = prevRotationKeyFrame.modelRotations[boxName]!!
                        box.rotation.set(currentQuat.i(), currentQuat.j(), currentQuat.k(), currentQuat.r())
                    } else if (prevRotationKeyFrame != null && nextRotationKeyFrame != null) { // Between keyframes
                        val currentQuat = Quaternion(0f, 0f, 0f, 0f)
                        currentQuat.slerp(
                            prevRotationKeyFrame.modelRotations[boxName]!!,
                            nextRotationKeyFrame.modelRotations[boxName]!!,
                            slerpProgress
                        )
                        box.rotation.set(currentQuat.i(), currentQuat.j(), currentQuat.k(), currentQuat.r())
                    } else { // There are no key frames
                        box.resetRotationQuaternion()
                    }

                    // Translations
                    val (prevTranslationsKeyFramePosition, prevTranslationKeyFrame) =
                        channel.getPreviousTranslationKeyFrameForBox(boxName, animCurrentFrame[channel.name]!!)
                    val (nextTranslationsKeyFramePosition, nextTranslationKeyFrame) =
                        channel.getNextTranslationKeyFrameForBox(boxName, animCurrentFrame[channel.name]!!)

                    var lerpProgress =
                        (currentFrame - prevTranslationsKeyFramePosition) / (nextTranslationsKeyFramePosition - prevTranslationsKeyFramePosition)
                    lerpProgress = lerpProgress.coerceAtMost(1f)

                    if (prevTranslationKeyFrame == null && nextTranslationKeyFrame != null) { // Before the first key frame
                        val startPosition = box.getRotationPointAsVector()
                        val endPosition = nextTranslationKeyFrame.modelTranslations[boxName]!!
                        val currentPosition = startPosition.copy()
                        currentPosition.interpolate(endPosition, lerpProgress)
                        box.setRotationPoint(currentPosition.x(), currentPosition.y(), currentPosition.z())
                    } else if (prevTranslationKeyFrame != null && nextTranslationKeyFrame == null) { // After the last keyframe
                        val currentPosition = prevTranslationKeyFrame.modelTranslations[boxName]!!
                        box.setRotationPoint(currentPosition.x(), currentPosition.y(), currentPosition.z())
                    } else if (prevTranslationKeyFrame != null && nextTranslationKeyFrame != null) { // Between keyframes
                        val startPosition = prevTranslationKeyFrame.modelTranslations[boxName]!!
                        val endPosition = nextTranslationKeyFrame.modelTranslations[boxName]!!
                        val currentPosition = startPosition.copy()
                        currentPosition.interpolate(endPosition, lerpProgress)
                        box.setRotationPoint(currentPosition.x(), currentPosition.y(), currentPosition.z())
                    } else { // There are no key frames
                        box.resetRotationPoint()
                    }
                } else {
                    (channel as CustomChannel).update(parts, this)
                }
            }
        }
    }
}